lib/rules/no-useless-assignment.js

```diff
@@ -0,0 +1,473 @@
+
+                        if (subsequentSegmentData.subsequentSegments.has(nextSegment)) {
+                            continue;
+                        }
+                        subsequentSegmentData.subsequentSegments.add(nextSegment);
+
+                        const assignmentInSegment = allAssignments.find(otherAssignment => otherAssignment.segment === nextSegment);
+
+                        if (!assignmentInSegment) {
+
+                            /*
+                             * Stores the next segment to explore.
+                             * If `assignmentInSegment` exists,
+                             * we are guarding it because we don't need to explore the next segment.
+                             */
+                            subsequentSegmentData.queueSegments.push(...nextSegment.nextSegments);
+                        }
+
+                        /** @type {SubsequentSegmentData} */
+                        const result = {
+                            segment: nextSegment,
+                            assignment: assignmentInSegment
+                        };
+
+                        subsequentSegmentData.results.push(result);
+                        yield result;
+                    }
+                }
+
+
+                const readReferences = targetAssignment.variable.references.filter(reference => reference.isRead());
+
+                if (!readReferences.length) {
+
+                    /*
+                     * It is not an unnecessary assignment, it is clearly an unnecessary (unused) variable
+                     * and should not be reported in the this rule. And it is reported by `no-unused-vars`.
+                     */
+                    return;
+                }
+
+                /**
+                 * Other assignment on the current segment and after current assignment.
+                 */
+                const otherAssignmentAfterTargetAssignment = allAssignments
+                    .find(assignment =>
+                        assignment !== targetAssignment &&
+                                assignment.segment === targetAssignment.segment &&
+                                targetAssignment.identifier.range[0] < assignment.identifier.range[0]);
+
+                for (const reference of readReferences) {
+
+                    /*
+                     * If the scope of the reference is outside the current code path scope,
+                     * we cannot track whether this assignment is not used.
+                     * For example, it can also be called asynchronously.
+                     */
+                    if (target.scope !== getCodePathStartScope(reference.from)) {
+                        return;
+                    }
+
+                    // Checks if it is used in the same segment as the target assignment.
+                    if (
+                        targetAssignment.identifier.range[0] < reference.identifier.range[0] &&
```

Comment:
I think this is a false negative due to `x` on the right side of `=` being considered as an usage of that assignment:

```js
/* eslint no-useless-assignment: "error" */

var x = 1;

x = x + 1; // false negative?

x = 5;

f(x);

```